### ISSUE TITLE
Success Criterion 3.1.2 - Language of Parts - Level AA

### ISSUE THREAD

**Issue by JJdeGroot** [w3c/matf#15]
Discussion:

<details>
<summary>MATF meeting on June 12, 2024</summary>

[Source](https://www.w3.org/2024/06/12-matf-minutes.html#t04)

JJ: (explaining the intent of 3.1.2)
… software replaces page
… section 508 this is excluded, also in the EN s301 549 - thi sis surprising because it can be done

JJ: content wise it can be challenging but that's the same for web stuff

<AlainVagner> To apply the related web page requirement for "Language of parts" to software would require the

<AlainVagner> marking-up of all text in all locations within the software. This would be impossible so an equivalent is

<AlainVagner> not included in this clause on software requirements

Alain: for the EN (reads the rationale why it is not covernd for software)

JJ: maybe it means that it is more difficult to achieve - not quite clear

JJ: why is it impossible?

<AlainVagner> I have checked in the GitLab of the EN 301 549, there is no issue about this SC.

JJ: possible in iOS and Androind, possibly more difficult in other Software environments
… no issue in the issue tracker for EN 301549 - maybe we should create one?
… was frst excluded in section 508, EN followed on

Gleidson: it is difficult to implement in the Android development environment - nothing documented

Gleidson: was referring to Jetpack Compose
</details>

Summary:

1. **Intent and Applicability of WCAG 3.1.2 to Software:**
    - **JJ:** Explained that `software` replaces `web page`, and noted that section 508 excludes this requirement, as does EN 301 549. Found this surprising because it can be implemented.
    - **JJ:** Acknowledged that content-wise, it can be challenging, but it is similar to web content challenges.

2. **Challenges of Implementing Language of Parts in Software:**
    - **Alain Vagner:** Highlighted that applying the web page requirement for "Language of parts" to software would necessitate marking up all text in all locations within the software, which is deemed impossible. This rationale is why it is not included in software requirements.
    - **Alain:** Read the rationale for its exclusion from the EN standards.
    - **JJ:** Questioned why it is considered "**impossible**" and noted that while it might be more difficult, it is not clear-cut.
    - **Alain Vagner:** Mentioned checking the GitLab of EN 301 549 and found no issue regarding this SC. Considered creating an issue in the EN 301 549 tracker due to its first exclusion in section 508 and subsequent exclusion in EN standards.
    - **JJ:** Suggested that it is possible in iOS and Android environments, though potentially more challenging in other software environments. 
    - **Gleidson:** Mentioned the difficulty of implementing this in the Android development environment, specifically referring to Jetpack Compose.

**Comment by arctouch-gleidsonramos** [comment]
I have run some tests with some Android engineers, and we are still not able to apply the same LocaleSpan or similar behavior in the Android Jetpack Compose. So, I have doubts about this being possible.
Could someone else try to do some tests, please?

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>27 November 2024</summary>

[Source](https://www.w3.org/2024/11/27-matf-minutes.html#106f)

~~~text
<dotjay> Apple have NLLanguageRecognizer which I suspect is responsible for triggering the automatic language detection we see in VoiceOvert: https://developer.apple.com/documentation/naturallanguage/identifying-the-language-in-text

<dotjay> Jamie ^^

dotjay can you add that link to the GH issue?

<dotjay> julianmka: Sure

thanks!

<Jamie> isn't https://developer.android.com/guide/topics/resources/app-languages addressing the issues for language of app for Android?

<Jamie> "Starting with Android Studio Giraffe and AGP 8.1, you can configure your app to support per-app language preferences automatically. "

<quintinb> Jamie the app needs to actually support that language - that's the problem. As a user, I can set the language to German, but it may default back to English because the app has no localisation for German

JJ: Summarizes discussion to date, including EN301 549's determination that meeting Language of Parts is impossible for software.

<quintinb> There are localisation files in Android development, if we could programmatically determine via a11y apis which one is loaded, that would solve it

<Jamie> can;t users can only set languages that are available in the OS settings for that app?

<quintinb> it's not about making *developers* lives easier, lol

JJ: Generally, it is possible to set language of parts, but not for all components/all places. Do we say it is not applicable?

ACTION: Decide how we apply this SC, given that it is technically possible, but would impose a burden on developers to use accessibility language supported strings everywhere. Could be added as metadata automatically by Google/Apple when loading localized resources.

<quintinb> I have to drop, sorry!

<JJ> Close the queue

Jamie: Teams I've worked with sometimes state that SCs that cannot be met is due to limitations of the technology. It is still on the dev team to track when things change, but allows a pass for times when developers cannot do something due to the platform technology.

<dotjay> julianmka: [w3c/matf#14 (comment)](https://github.com/w3c/matf/issues/14#issuecomment-2504101176)

Jamie: Do we need to add a note about hybrid/web view content?

ACTION: Add note about `lang` tag for hybrid/webview content to pass 3.1.1 and 3.1.2

JJ: yes, web view content should follow web guidance.

julianmka: Also need to consider webview content rendered within a regular view, not just an in-app browser.
~~~
</details>

ACTION: Decide how we apply this SC, given that it is technically possible, but would impose a burden on developers to use accessibility language supported strings everywhere. Could be added as metadata automatically by Google/Apple when loading localized resources.

ACTION: Add note about `lang` tag for hybrid/webview content to pass 3.1.1 and 3.1.2

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>9 April 2025</summary>

[Source](https://www.w3.org/2025/04/09-matf-minutes.html#e581)

~~~text
<JJ> https://w3c.github.io/matf/#success-criterion-3-1-2-language-of-parts

<JJ> [w3c/matf#15](https://github.com/w3c/matf/issues/15)

JJ: Summarized background on 3.1.2 about feasibility of applying this SC. It is technically possible, but could be a burden for developers. Also, dealing with web content (use of <lang> passes the SC for web views).

<pauljadam> content can already describe everything

<pauljadam> it means it fails WCAG as a defect in the platform the platform developer must fix

JJ: WCAG2ICT replaced "content" with "non-web document or software" and added a note that some technologies don't support this and "may not be able to meet" this SC. Is "may not be able to meet" an exemption or a failure?

ACTION: Find out what "may not able to meet" means for 3.1.1 and 3.1.2

JJ: Android and iOS do allow this, though we need to map out other x-platform frameworks. But, is it a burden on developers? It seems to me that it may not be much worse on mobile than on websites.

<pauljadam> iOS SwiftUI a11y docs on Language of Parts https://github.com/cvs-health/ios-swiftui-accessibility-techniques/blob/main/iOSswiftUIa11yTechniques/Documentation/Language.md

AlainVagner: Not sure if it's a burden, but coming from a country with more than one national language, we often mix languages and it's a burden when you don't indicate the language. Feels like a loophole for mobile vs. what's available for web.
… I have a feeling it's not much more complex than indicating language on the web. There may be some cases where it's not possible, and for me that would be an exemption.

<pauljadam> It is the case that you can set the language for strings in a language different than the page. If something does not work then it's a platform defect failure that apple or google would need to fix. No exception to WCAG needed!

JJ: On Mobile, components must accept attributed strings or localized span.

<JJ> pauljadam > imagine a component that does not take NSAttributedString as argument, you won't be able to pass a localized string, even if the OS does support it (the component does not)

<pauljadam> Are well saying that "On Android you can't set the heading level so therefore you are exempt"? Because I would not provide an exemption, It should just be a generally understood rule that If you can meet WCAG here then you must, if you can't then your platform has a defect and apple or google need to fix it.

Jon_Gibbins: This reminds me of WCAG 1.0 days. Speech engines will detect a change of language on platforms, which doesn't mean we should not also give language info attached to strings. We're not just writing guidelines for Android and iOS, but for mobile as an environment. We should aim for the spirit of WCAG, and when the platform doesn't

support it, there's an exemption. But the work is on devs to determine what's possible.

<pauljadam> Man I wish we could use Slack for meetings and chats so I could edit my typos 🤪

JJ: I agree, we could use a note that there would be an exception for cases where the chosen platform cannot support the language.

<Jon_Gibbins> JJ: Agreed

JJ: We could take an approach like with 1.1.1 where the SC applies as written, but include a note that not all platforms support lang of parts.

pauljadam: I don't agree with the label and image not being possible on all platforms. I'm not a fan of adding notes that say "this isn't possible", what happens when the platform evolves and improves? It should be understood that devs need to follow WCAG and if they can't, that's a platform defect that Google and Apple should fix.

<Jon_Gibbins> PaulJAdam: Agreed on notes. We can capture platform specific detail on any supporting understanding documentation.

JJ: Part of our work as MATF is to point out instances where there may be technological barriers. The note for 1.1.1 is to address expected behavior where someone might focus on an image, the caption below would be read.

<pauljadam> what about when a new mobile OS comes out, ChatGPT mobileOS

<Jon_Gibbins> For clarity, my point about WCAG 1 was in having a common phrase that we use to refer to situations where we know that certain platforms are currently lacking *without* refering to specific platforms in this guidance document of ours translating WCAG2ICT.

JJ: During the last CFC, a question of scope came out and we do need to clarify the platforms we're considering in this doc.

As for updates, the major platforms get updated once a year, it seems reasonable to update WCAG2ICT once a year if needed.
… We could make a distinction between general guidance and more specific notes that apply to Android or iOS

<Jon_Gibbins> For reference: https://www.w3.org/WAI/Resources/WAI-UA-Support

<Jon_Gibbins> Joe_Humbert +1

Joe_Humbert: I think we do need to have notes, but based on CFC feedback, we should try to make notes as generic as possible so that they could cover other devices/OSs besides Android and iOS. We've done a decent job where notes do apply to Android and iOS but they could apply to other OSs.

JJ: Maybe we should keep track of which notes target which platforms. LIke the 1.1.1 note mostly applies to iOS.

<pauljadam> That note makes me worry that a developer would see that and say "Oh we can't make labels accessible but have an exception".

<pauljadam> The note is not accurate, I can code an example with a programmatic label.

Joe_Humbert: Maybe we include an example as part of the note.

JJ: This might help for OS-level considerations, like known examples, since Understanding docs are off the table for the moment.

<pauljadam> I think talking about what is possible can be done in Mobile Techniques rather than the notes that say what is Not Possible.

JJ: Re Paul's comment about the label, it's perhaps not the best example because there are workarounds. 3.1.1 would be a better example.

ACTION: Mention platforms in the notes, e.g. it's known that on Android certain functionality is not available, while keeping the note generically applicable to more mobile operating systems

Jon_Gibbins: I agree with Joe and Paul's comments about making this doc cover as many bases as we can, not being overly specific about platforms. There's an opportunity and a danger: Opportunity to educate developers, but it's not the doc's job to provide specific platform techniques.
… To the 1.1.1 note about labels, we have to be super careful about vocabulary we're using because it's WCAG terminology.

<JJ> Jon_Gibbins: We have to be very careful with WCAG (web) terminology in our mobile application guidance to avoid confusion / wrong expectations

Jon_Gibbins: We can link off to existing resources on techniques like Appt or the iOS techniques repo. We have to be clear that in this example an accessible name is a label, not a label-for. Or we can add a note that we are talking about a label and it doesn't mean accessibilityLabel or Content DEscription.

<pauljadam> If WCAG can have PDF Techniques, Silverlight techniques. I'm not sure why we can't have Mobile Techniques 🤷‍♂️

<JJ> Because those techniques can be used on the Web, and mobile apps are 'Non-Web' :/

<pauljadam> Mobile apps live on the web as far as end users think of it. HTML is the web and Mobile is the web.

Jon_Gibbins: In our notes, we're not saying that an SC doesn't apply, but that there may be issues in applying it. We should be careful with what we put in notes.

<Tanya> Was is ever considered to add a note something like "if there are constraints or limitations set by the operating system, the implementation should follow the best possible or closest alternative that meets the intent."?

JJ: We have a wide audience -- accessibility professionals, developers who are familiar with WCAG, developers who are new to WCAG. Also, I want to revisit the expand/collapse mechanism, will people actually read their content?

<Jon_Gibbins> pauljadam: WCAG does not have techniques in it. The supporting Understanding documents do.
~~~
</details>

- ACTION: Find out what "may not able to meet" means for 3.1.1 and 3.1.2
- ACTION: Mention platforms in the notes, e.g. it's known that on Android certain functionality is not available, while keeping the note generically applicable to more mobile operating systems

**Comment by mitchellevan** [comment]
> section 508 excludes this requirement, as does EN 301 549

For what it's worth, Section 508 actually does include SC 3.1.2 for non-web software. [E207.2 WCAG Conformance](https://www.access-board.gov/ict/#E207.2)

This difference between Section 508 and EN 301 549 is not necessarily wrong, because Section 508 offers the possibility of exempting platform-specific limitations under [E202.6 Undue Burden or Fundamental Alteration](https://www.access-board.gov/ict/#E202.6).

**Comment by jha11y** [comment]
Maybe these resources are helpful?

https://developer.android.com/guide/topics/resources/multilingual-support#specify-the-languages-your-app-supports

https://developer.android.com/training/basics/supporting-devices/languages#app-languages

**Comment by JJdeGroot** [comment]
[Discussed in MATF meeting on 1 April 2026](https://www.w3.org/2026/04/01-matf-minutes.html#2da6)

<details>
<summary>Minutes</summary>

~~~text
JJ: feedback on PR most people voted on this one and it's good to go

Tanya asking how to proceed on lang of parts

<JJ> Existing draft: [w3c/matf#278](https://github.com/w3c/matf/pull/278)

JJ: netlify and markdown technical parts may need updates

JJ: more deployments once the PRs are finished soon

shoobe01: issues with netlify links, and talking about workaround to read HTML to approve

JJ: may need to explain about how to install NPM dependences to generate HTML

JJ: will be adding guidance about GitHub how to review branch with work shown in your browser
~~~
</details>
